powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/internal/ValidationAspect.java

```diff
@@ -118,131 +126,215 @@ public Object around(ProceedingJoinPoint pjp,
-        		if (result instanceof APIGatewayProxyResponseEvent) {
-        			APIGatewayProxyResponseEvent response = (APIGatewayProxyResponseEvent) result;
-        			overridenResponse = validateAPIGatewayProxyBody(response.getBody(), outboundJsonSchema, response.getHeaders(),
-                response.getMultiValueHeaders());
-        		} else if (result instanceof APIGatewayV2HTTPResponse) {
-        			APIGatewayV2HTTPResponse response = (APIGatewayV2HTTPResponse) result;
-              overridenResponse = validateAPIGatewayV2HTTPBody(response.getBody(), outboundJsonSchema, response.getHeaders(),
-                response.getMultiValueHeaders());
-              // all type of below responses will throw an exception if validation fails
-        		} else if (result instanceof APIGatewayV2WebSocketResponse) {
-        			APIGatewayV2WebSocketResponse response = (APIGatewayV2WebSocketResponse) result;
-        			validate(response.getBody(), outboundJsonSchema);
-        		} else if (result instanceof ApplicationLoadBalancerResponseEvent) {
-        			ApplicationLoadBalancerResponseEvent response = (ApplicationLoadBalancerResponseEvent) result;
-        			validate(response.getBody(), outboundJsonSchema);
-        		} else if (result instanceof KinesisAnalyticsInputPreprocessingResponse) {
-        			KinesisAnalyticsInputPreprocessingResponse response =
-        					(KinesisAnalyticsInputPreprocessingResponse) result;
-        			response.getRecords().forEach(record -> validate(decode(record.getData()), outboundJsonSchema));
-        		} else {
-        			LOG.warn("Unhandled response type {}, please use the 'envelope' parameter to specify what to validate",
-        					result.getClass().getName());
-        		}
-        		
-        		if (overridenResponse != null) {
-        			result = overridenResponse;
-        			LOG.error("API response failed outbound schema validation.");
-        		}
-        	}
-        } 
 
         return result;
     }
-    
+
+    /**
+     * Validate each Kinesis record body. If an error occurs, do not fail the whole batch but only add invalid items in BatchItemFailure.
+     * @param records Kinesis records
+     * @param inboundJsonSchema validation schema
+     * @return the stream response with items in failure
+     */
+    private StreamsEventResponse validateKinesisEventRecords(List<KinesisEvent.KinesisEventRecord> records,
+                                                             JsonSchema inboundJsonSchema) {
+        StreamsEventResponse response = StreamsEventResponse.builder().withBatchItemFailures(new ArrayList<>()).build();
+
+        ListIterator<KinesisEvent.KinesisEventRecord> listIterator = records.listIterator(); // using iterator to remove while browsing
+        while (listIterator.hasNext()) {
+            KinesisEvent.KinesisEventRecord eventRecord = listIterator.next();
+            try {
+                validate(decode(eventRecord.getKinesis().getData()), inboundJsonSchema);
```

Comment:
We're going to decode the record bodies twice, right? Once here, and once again on the user's handler (for SQS too)

Reply:
yes for the valid items (as invalid ones are removed), which means when messages are mostly valid, it can slow down the global process... any idea to improve this ?

Reply:
Nothing that feels clean. I suggest we just put a comment on so it's obvious when reading the code this happens. Then when we stumble over it in the future something clean might have popped up.